garage/np/algos/batch_polopt.py

```diff
@@ -115,6 +118,155 @@ def log_diagnostics(self, paths):
+
+        if hasattr(self.baseline, 'predict_n'):
+            all_path_baselines = self.baseline.predict_n(paths)
+        else:
+            all_path_baselines = [
+                self.baseline.predict(path) for path in paths
+            ]
+
+        for idx, path in enumerate(paths):
+            path_baselines = np.append(all_path_baselines[idx], 0)
+            deltas = path['rewards'] + \
+                self.discount * path_baselines[1:] - path_baselines[:-1]
+            path['advantages'] = special.discount_cumsum(
+                deltas, self.discount * self.gae_lambda)
+            path['returns'] = special.discount_cumsum(path['rewards'],
+                                                      self.discount)
+            baselines.append(path_baselines[:-1])
+            returns.append(path['returns'])
+
+        ev = special.explained_variance_1d(
+            np.concatenate(baselines), np.concatenate(returns))
+
+        if not self.policy.recurrent:
+            observations = tensor_utils.concat_tensor_list(
+                [path['observations'] for path in paths])
+            actions = tensor_utils.concat_tensor_list(
+                [path['actions'] for path in paths])
+            rewards = tensor_utils.concat_tensor_list(
+                [path['rewards'] for path in paths])
+            returns = tensor_utils.concat_tensor_list(
+                [path['returns'] for path in paths])
+            advantages = tensor_utils.concat_tensor_list(
+                [path['advantages'] for path in paths])
+            env_infos = tensor_utils.concat_tensor_dict_list(
+                [path['env_infos'] for path in paths])
+            agent_infos = tensor_utils.concat_tensor_dict_list(
+                [path['agent_infos'] for path in paths])
+
+            if self.center_adv:
+                advantages = utils.center_advantages(advantages)
+
+            if self.positive_adv:
+                advantages = utils.shift_advantages_to_positive(advantages)
+
+            average_discounted_return = \
+                np.mean([path['returns'][0] for path in paths])
+
+            undiscounted_returns = [sum(path['rewards']) for path in paths]
+
+            ent = np.mean(self.policy.distribution.entropy(agent_infos))
+
+            samples_data = dict(
+                observations=observations,
+                actions=actions,
+                rewards=rewards,
+                returns=returns,
+                advantages=advantages,
+                env_infos=env_infos,
+                agent_infos=agent_infos,
+                paths=paths,
+            )
+        else:
+            max_path_length = max([len(path['advantages']) for path in paths])
+
+            # make all paths the same length (pad extra advantages with 0)
+            obs = [path['observations'] for path in paths]
+            obs = tensor_utils.pad_tensor_n(obs, max_path_length)
+
+            if self.center_adv:
+                raw_adv = np.concatenate(
```

Comment:
Why not use `utils.center_adv` too?

Reply:
some of this is implemented in tensorflow for NPO now, and parts of process_samples may be redundant.